pycbc/events/stat.py

```diff
@@ -1758,6 +1757,155 @@ def coinc_lim_for_thresh(self, s, thresh, limifo,
+                               "attribute called {dname}-kde_file !")
+        self.kde_by_tid = {}
+        for dname in self.data_name:
+            self.kde_by_tid[dname] = self.assign_kdes(dname)
+            #print('self.kde_by_tid=', self.kde_by_tid[dname])
+        self.curr_tnum = None       
+ 
+    def assign_kdes(self, dname):
+        """
+        Extract fits from fit files
+
+        Parameters
+        -----------
+        dname: str
+            signal or template.
+
+        Returns
+        -------
+        rate_dict: dict
+            A dictionary containing kde values for signal and template files
+
+        """
+        kde_file = self.files[dname+'-kde_file']
+        template_id = kde_file['template_id'][:]
+        tid_sort = numpy.argsort(template_id)
+        kde_by_tid_dict = {}
+        kde_by_tid_dict[dname+'_data_kde'] = \
+            kde_file['data_kde'][:][tid_sort]
+        print('self.kde_by_tid_dict', kde_by_tid_dict)
+        return kde_by_tid_dict
+
+    def rank_stat_coinc(self, stats, shift, to_shift):
+        """
+        Calculate the normalized log rate density of signals via lookup
+
+        This calls back to the Parent class and then applies the ratio_kde
+        weighting factor.
+
+        Parameters
+        ----------
+        stats: list of dicts giving single-ifo quantities, ordered as
+            self.ifos
+        shift: numpy array of float, size of the time shift vector for each
+            coinc to be ranked
+        to_shift: list of int, multiple of the time shift to apply ordered
+            as self.ifos
+
+        Returns
+        -------
+        value: log of coinc signal rate density for the given single-ifo
+            triggers and time shifts
+        """
+        logr_s = ExpFitFgBgNormStatistic.logsignalrate(self, stats,
+                                                       shift, to_shift)
+        signal_kde = self.kde_by_tid["signal"]
+        template_kde = self.kde_by_tid["template"]
+        logr_s += [numpy.log(i/j) for i, j in zip(signal_kde, template_kde)]
+        print('logr_s=', logr_s)
+        return logr_s
+
+    def single(self, trigs):
+        """
+        Calculate the necessary single detector information
+
+        In this case the ranking rescaled (see the lognoiserate method here)
+        with the phase, end time, sigma, SNR, template_id and the
+        benchmark_logvol values added in.
+
+        Parameters
+        ----------
+        trigs: dict of numpy.ndarrays, h5py group (or similar dict-like object)
+            Dictionary-like object holding single detector trigger information.
+
+        Returns
+        -------
+        numpy.ndarray
+            The array of single detector values
+        """
+        try:
+            tnum = trigs.template_num
+        except AttributeError:
+            tnum = trigs['template_id']
+        self.curr_tnum = self.kde_by_tid[tnum]
+        print(self.curr_tnum)
```

Comment:
Again we will not want to have print statements in the code when used in production.

Reply:
What do you want to do with the curr_tnum variable? You are not using it for anything.